cmd/node/multi_node.go

```diff
@@ -0,0 +1,260 @@
+
+	swarm := net.NewNode()
+	dbStorepath := storePath
+
+	hareOracle := oracle.NewLocalOracle(rolacle, 5, nodeID)
+	hareOracle.Register(true, pub.String())
+
+	postClient, err := activation.NewPostClient(&smApp.Config.POST, util.Hex2Bytes(nodeID.Key))
+	if err != nil {
+		return nil, err
+	}
+
+	err = smApp.initServices(nodeID, swarm, dbStorepath, edSgn, false, hareOracle, uint32(smApp.Config.LayerAvgSize), postClient, poetClient, vrfSigner, uint16(smApp.Config.LayersPerEpoch), clock)
+	if err != nil {
+		return nil, err
+	}
+	smApp.setupGenesis()
+
+	return smApp, err
+}
+
+// Starts the run of a number of nodes, running in process consensus between them.
+// this also runs a single transaction between the nodes.
+func StartMultiNode(numOfinstances, layerAvgSize int, runUntilLayer uint32, dbPath string) {
+
+	runTillLayer := runUntilLayer
+
+	cfg := getTestDefaultConfig()
+	cfg.LayerAvgSize = layerAvgSize
+	numOfInstances := numOfinstances
+
+	dst := types.BytesToAddress([]byte{0x02})
+	acc1Signer, err := signing.NewEdSignerFromBuffer(util.FromHex(apiCfg.Account1Private))
+	if err != nil {
+		log.Panic("Could not build ed signer err=%v", err)
+	}
+	tx, err := mesh.NewSignedTx(0, dst, 10, 1, 1, acc1Signer)
+	if err != nil {
+		log.Panic("panicked creating signed tx err=%v", err)
+	}
+	txbytes, _ := types.InterfaceToBytes(tx)
+	path := dbPath + time.Now().Format("2006-01-02T15:04:05-0700")
+
+	genesisTime := time.Now().Add(20 * time.Second).Format(time.RFC3339)
+
+	poetClient, err := NewRPCPoetHarnessClient()
+	if err != nil {
+		log.Panic("failed creating poet client harness: %v", err)
+	}
+	defer poetClient.Teardown(true)
+
+	rolacle := eligibility.New()
+	rng := BLS381.DefaultSeed()
+	gTime, err := time.Parse(time.RFC3339, genesisTime)
+	if err != nil {
+		log.Error("cannot parse genesis time %v", err)
+	}
+	pubsubAddr := "tcp://localhost:56565"
+	events.InitializeEventPubsub(pubsubAddr)
+	clock := timesync.NewManualClock(gTime)
+
+	apps := make([]*SpacemeshApp, 0, numOfInstances)
+	name := 'a'
+	for i := 0; i < numOfInstances; i++ {
+		dbStorepath := path + string(name)
+		smApp, err := InitSingleInstance(*cfg, i, genesisTime, rng, dbStorepath, rolacle, poetClient, true, clock)
+		if err != nil {
+			log.Error("cannot run multi node %v", err)
+			return
+		}
+		apps = append(apps, smApp)
+		name++
+	}
+
+	eventDb := collector.NewMemoryCollector()
+	collect := collector.NewCollector(eventDb, pubsubAddr)
+	for _, a := range apps {
+		a.startServices()
+	}
+	collect.Start(false)
+	ActivateGrpcServer(apps[0])
+
+	if err := poetClient.Start("127.0.0.1:9091"); err != nil {
+		log.Panic("failed to start poet server: %v", err)
+	}
+
+	//startInLayer := 5 // delayed pod will start in this layer
+	defer GracefulShutdown(apps)
+
+	_ = apps[0].P2P.Broadcast(miner.IncomingTxProtocol, txbytes)
+	timeout := time.After(time.Duration(runTillLayer*60) * time.Second)
+
+	//stickyClientsDone := 0
+	startLayer := time.Now()
+	clock.Tick()
+loop:
```

Comment:
I'd refactor this loop outside of this function

Reply:
what is the purpose?

Reply:
the function's name is StartMultiNode.. this is not what you're doing in this loop